#Data Cleaning - YouTube Retention Project

In [1]:
import pandas as pd 
import numpy as np

Loading data

In [2]:
#Loading dataset

df = pd.read_csv("../data/raw/USvideos.csv")
df.head()

,video_id,trending_date,title,channel_title,category_id,publish_time,tags,views,likes,dislikes,comment_count,thumbnail_link,comments_disabled,ratings_disabled,video_error_or_removed,description
0,2kyS6SvSYSE,17.14.11,WE WANT TO TALK ABOUT OUR MARRIAGE,CaseyNeistat,22,2017-11-13T17:13:01.000Z,SHANtell martin,748374,57527,2966,15954,https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg,False,False,False,SHANTELL'S CHANNEL - https://www.youtube.com/s...
1,1ZAPwfrtAFY,17.14.11,The Trump Presidency: Last Week Tonight with J...,LastWeekTonight,24,2017-11-13T07:30:00.000Z,"last week tonight trump presidency|""last week ...",2418783,97185,6146,12703,https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg,False,False,False,"One year after the presidential election, John..."
2,5qpjK5DgCt4,17.14.11,"Racist Superman | Rudy Mancuso, King Bach & Le...",Rudy Mancuso,23,2017-11-12T19:05:24.000Z,"racist superman|""rudy""|""mancuso""|""king""|""bach""...",3191434,146033,5339,8181,https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg,False,False,False,WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...
3,puqaWrEC7tY,17.14.11,Nickelback Lyrics: Real or Fake?,Good Mythical Morning,24,2017-11-13T11:00:04.000Z,"rhett and link|""gmm""|""good mythical morning""|""...",343168,10172,666,2146,https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg,False,False,False,Today we find out if Link is a Nickelback amat...
4,d380meD0W0M,17.14.11,I Dare You: GOING BALD!?,nigahiga,24,2017-11-12T18:01:41.000Z,"ryan|""higa""|""higatv""|""nigahiga""|""i dare you""|""...",2095731,132235,1989,17518,https://i.ytimg.com/vi/d380meD0W0M/default.jpg,False,False,False,I know it's been a while since we did this sho...


##Objective:
Clean and prepare the dataset to ensure reliable analysis and modeling.


In [6]:
df.shape


(40949, 16)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40949 entries, 0 to 40948
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   video_id                40949 non-null  object
 1   trending_date           40949 non-null  object
 2   title                   40949 non-null  object
 3   channel_title           40949 non-null  object
 4   category_id             40949 non-null  int64 
 5   publish_time            40949 non-null  object
 6   tags                    40949 non-null  object
 7   views                   40949 non-null  int64 
 8   likes                   40949 non-null  int64 
 9   dislikes                40949 non-null  int64 
 10  comment_count           40949 non-null  int64 
 11  thumbnail_link          40949 non-null  object
 12  comments_disabled       40949 non-null  bool  
 13  ratings_disabled        40949 non-null  bool  
 14  video_error_or_removed  40949 non-null  bool  
 15  de

##Data Overview
- Dataset contains ~41K rows and 16 columns
- Includes engagement metrics (views, likes, comments)
- publish_time requires datatype conversion
- Data is initially skewed with extreme values

In [8]:
df.describe()

,category_id,views,likes,dislikes,comment_count
count,40949.000000,4.094900e+04,4.094900e+04,4.094900e+04,4.094900e+04
mean,19.972429,2.360785e+06,7.426670e+04,3.711401e+03,8.446804e+03
std,7.568327,7.394114e+06,2.288853e+05,2.902971e+04,3.743049e+04
min,1.000000,5.490000e+02,0.000000e+00,0.000000e+00,0.000000e+00
25%,17.000000,2.423290e+05,5.424000e+03,2.020000e+02,6.140000e+02
50%,24.000000,6.818610e+05,1.809100e+04,6.310000e+02,1.856000e+03
75%,25.000000,1.823157e+06,5.541700e+04,1.938000e+03,5.755000e+03
max,43.000000,2.252119e+08,5.613827e+06,1.674420e+06,1.361580e+06


In [9]:
# Checking missing values

df.isnull().sum()

video_id                    0
trending_date               0
title                       0
channel_title               0
category_id                 0
publish_time                0
tags                        0
views                       0
likes                       0
dislikes                    0
comment_count               0
thumbnail_link              0
comments_disabled           0
ratings_disabled            0
video_error_or_removed      0
description               570
dtype: int64

##Missing Values Handling
- description column had ~570 missing values
- These rows were removed since text is required for sentiment analysis

In [11]:
#handling missing values
df = df.dropna(subset=['description'])

In [13]:
# remove duplicates
df = df.drop_duplicates()

In [14]:
# remove invalid videos
df = df[df['video_error_or_removed'] == False]

In [15]:
#convert datatypes

df['publish_time'] = pd.to_datetime(df['publish_time'], errors='coerce')
df = df.dropna(subset=['publish_time'])

## Cleaning Steps Performed
- Removed duplicate records
- Dropped rows with missing descriptions
- Filtered out invalid/removed videos
- Converted `publish_time` to datetime format

In [16]:
# handling outliers

upper_limit = df['views'].quantile(0.99)
df = df[df['views'] < upper_limit]

## Outlier Handling
- Top 1% of videos by views were removed
- This prevents extreme viral videos from biasing the analysis
- Ensures more realistic performance distribution

checking

In [17]:
df.shape

(39907, 16)

In [18]:
df.isnull().sum()

video_id                  0
trending_date             0
title                     0
channel_title             0
category_id               0
publish_time              0
tags                      0
views                     0
likes                     0
dislikes                  0
comment_count             0
thumbnail_link            0
comments_disabled         0
ratings_disabled          0
video_error_or_removed    0
description               0
dtype: int64

In [19]:
df.describe()

,category_id,views,likes,dislikes,comment_count
count,39907.000000,3.990700e+04,3.990700e+04,39907.000000,39907.000000
mean,20.015461,1.801583e+06,6.040335e+04,2632.014834,6686.532187
std,7.566812,3.341707e+06,1.392797e+05,10964.817767,19994.391634
min,1.000000,5.490000e+02,0.000000e+00,0.000000,0.000000
25%,17.000000,2.434850e+05,5.524000e+03,202.000000,616.000000
50%,24.000000,6.760240e+05,1.795300e+04,624.000000,1841.000000
75%,25.000000,1.769134e+06,5.355050e+04,1862.500000,5601.000000
max,43.000000,2.994356e+07,2.700800e+06,497847.000000,658130.000000


In [20]:
df.to_csv("../data/processed/cleaned_videos.csv", index=False)

In [21]:
df[['views']].describe(percentiles=[0.9, 0.95,0.99])

,views
count,3.990700e+04
mean,1.801583e+06
std,3.341707e+06
min,5.490000e+02
50%,6.760240e+05
90%,4.216228e+06
95%,7.738310e+06
99%,1.827655e+07
max,2.994356e+07


## Final Dataset Summary
- Final dataset contains ~39K rows
- No missing values remain
- Data is clean and ready for feature engineering
- Distribution is still skewed but controlled